src/libsyntax/parse/parser.rs

```diff
@@ -5232,23 +5233,23 @@ impl<'a> Parser<'a> {
             }
             let mut err = self.diagnostic().struct_span_err(id_sp,
                 "cannot declare a new module at this location");
-            let this_module = match self.directory.path.file_name() {
-                Some(file_name) => file_name.to_str().unwrap().to_owned(),
-                None => self.root_module_name.as_ref().unwrap().clone(),
-            };
-            err.span_note(id_sp,
-                          &format!("maybe move this module `{0}` to its own directory \
-                                     via `{0}/mod.rs`",
-                                    this_module));
+            if id_sp != syntax_pos::DUMMY_SP {
+                let mut full_path = PathBuf::from(self.sess.codemap().span_to_filename(id_sp));
+                let full_path_copy = full_path.clone();
+                let stem = full_path_copy.file_stem().unwrap_or(OsStr::new(""));
+                full_path.set_file_name(stem);
+                full_path.push("mod.rs");
+                err.span_note(id_sp,
+                              &format!("maybe move this module `{0}` to its own directory \
+                                        via `{0}`", full_path.to_str().unwrap_or("")));
```

Comment:
Don't we want two paths here, e.g. ``maybe move this module `foo/bar.rs` to its own directory via `foo/bar/mod.rs` ``?

Also, instead of `unwrap_or("")`, I would use `to_string_lossy` or just not emit the note if `to_str` is `None`.